pkg/test/framework/driver.go

```diff
@@ -12,153 +12,117 @@
-	"istio.io/istio/pkg/test/dependency"
-	"istio.io/istio/pkg/test/environment"
-	"istio.io/istio/pkg/test/internal"
-	"istio.io/istio/pkg/test/local"
+	"istio.io/istio/pkg/test/framework/dependency"
+	"istio.io/istio/pkg/test/framework/environment"
+	"istio.io/istio/pkg/test/framework/environments/kubernetes"
+	"istio.io/istio/pkg/test/framework/environments/local"
+	"istio.io/istio/pkg/test/framework/internal"
+	"istio.io/istio/pkg/test/framework/settings"
 )
 
+// internal state for the driver.
+type state int
+
 const (
-	maxTestIDLength = 30
-)
+	// Runtime states for the driver
+
+	// The driver is freshly created. It hasn't run yet.
+	created state = iota
 
-var scope = log.RegisterScope("testframework", "General scope for the test framework", 0)
+	// The driver is currently running tests.
+	running
+
+	// the driver has completed running tests.
+	completed
+)
 
 type driver struct {
 	lock sync.Mutex
 
-	args *Args
-	ctx  *internal.TestContext
+	settings *settings.Settings
+	ctx      *internal.TestContext
 
 	// The names of the tests that we've encountered so far.
 	testNames map[string]struct{}
 
-	running bool
+	state state
 }
 
-var _ Interface = &driver{}
-
 // New returns a new driver instance.
-func New() Interface {
+func newDriver() *driver {
 	return &driver{
 		testNames: make(map[string]struct{}),
+		state:     created,
 	}
 }
 
-// Initialize implements same-named Interface method.
-func (d *driver) Initialize(a *Args) error {
-	scope.Debugf("Enter: driver.Initialize (%s)", a.TestID)
+// Run implements same-named Driver method.
+func (d *driver) Run(testID string, m *testing.M) (int, error) {
 	d.lock.Lock()
-	defer d.lock.Unlock()
-
-	if d.ctx != nil {
-		return errors.New("test driver is already initialized")
-	}
-
-	// Make a copy of the args
-	args := &(*a)
-	if err := args.Validate(); err != nil {
-		return err
-	}
-
-	// Initialize the environment.
-	var env internal.Environment
-	switch a.Environment {
-	case EnvLocal:
-		env = local.NewEnvironment()
-	case EnvKube:
-		env = cluster.NewEnvironment()
-	default:
-		return fmt.Errorf("unrecognized environment: %s", a.Environment)
-	}
-
-	// Create the context, but do not attach it to the member fields before initializing.
-	ctx := internal.NewTestContext(
-		args.TestID,
-		generateRunID(args.TestID),
-		a.WorkDir,
-		a.Hub,
-		a.Tag,
-		a.KubeConfig,
-		env)
-
-	if err := env.Initialize(ctx); err != nil {
-		return err
+	// do not defer unlock. We will explicitly unlock before starting test runs.
+	if d.state != created {
+		d.lock.Unlock()
+		return -1, fmt.Errorf("driver.Run must be called only once")
 	}
+	d.state = running
 
-	d.ctx = ctx
-	d.args = args
-
-	return nil
-}
-
-// Run implements same-named Interface method.
-func (d *driver) Run() int {
-	scope.Debugf("Enter: driver.Run (%s)", d.ctx.TestID())
-	d.lock.Lock()
-
-	// If context is not set, then we're not initialized.
-	if d.ctx == nil {
+	s, err := settings.Construct(testID)
```

Comment:
Seems we use the pattern `newXXX` elsewhere ... should this be `settings.NewSettings(testID)`?

Reply:
I'm not sure the distinction of `new` vs `construct` is very clear.  I understand that it's a slightly advanced form of `new`, but it's still fundamentally a `new` function.  I'd favor consistency rather than introducing new terms that are only understood by a few developers.